firedrake/cofunction.py

```diff
@@ -79,6 +79,18 @@ def __del__(self):
         if hasattr(self, "_comm"):
             mpi.decref(self._comm)
 
+    def __call__(self, primal):
+        r"""Compute the duality pairing, this is to evaluate this
+        :class:`.Cofunction` at a :class:`.Function`, equivalent to taking the
+        dot product of the vectors of coefficients of the dual and primal
+        arguments.
+
+        :arg primal: the :class:`.Function` to be paired with this :class:`.Cofunction`.
+
+        :returns: the scalar result of the duality pairing.
+        """
+        return firedrake.assemble(ufl.action(self, primal))
```

Comment:
I feel like just returning the symbolic object would facilitate things like differentiation or dual evaluation, especially with the upcoming `Interpolate`. For example, the dual evaluation when primal lives in a different space would just write:

```
c(Interpolate(f, V2))
```

where `c` is a `Cofunction` in `V2*` and f a `Function` in `V1`. Likewise, differentiation could be done seamlessly while accounting for the `Interpolate`.

I guess we could have a keyword argument for assembly, but then I don't think there is a real gain in writing something like:

```
c(f, assemble=True)
```

instead of

```
assemble(c(f))
```